### PR TITLE
ARROW-11719: [Rust][Datafusion] support creating memory table with merged schema

### PR DESCRIPTION
**Pull request by houqp**
* Added `contains` method for `arrow::datatypes::Schema` and
`arrow::datatypes::Field`
* Relax batch schema validation using `contains` check when creating a
MemTable in datafusion

### CODE REVIEW

File: rust/arrow/src/datatypes/schema.rs [hunk 1]
```diff
@@ -279,6 +279,37 @@ impl Schema {
             )),
         }
     }
+
+    /// Check to see if `self` is a superset of `other` schema Here are the comparision rules:
+    ///
+    /// * for every field `f` in other, the field in self with corresponding index should be a
```
Comment (alamb) [inline rust/arrow/src/datatypes/schema.rs:285]: 👍  thank you for the clear comments

File: rust/datafusion/src/datasource/memory.rs [hunk 2]
```diff
@@ -365,4 +366,59 @@ mod tests {
...
     }
+
+    #[tokio::test]
+    async fn test_merged_schema() -> Result<()> {
+        let mut metadata = HashMap::new();
+        metadata.insert("foo".to_string(), "bar".to_string());
+
+        let schema1 = Schema::new_with_metadata(
+            vec![
+                Field::new("a", DataType::Int32, false),
```
Comment (alamb) [inline rust/datafusion/src/datasource/memory.rs:377]: I think this test should also include different numbers of fields in the two record batches (e.g. remove "c" from the first batch and remove "b" from the second)

**Review by alamb — COMMENTED** [review]
This is a cool change @houqp  -- thanks. 

I am not sure if DataFusion will work with different schemas without some additional modifications. Specifically, when the schemas are actually subsets of each other with different numbers of columns -- here is what I came up with: https://github.com/influxdata/influxdb_iox/blob/main/query/src/provider/adapter.rs#L44-L70

I think the `contains` logic makes sense, and is actually quite interesting -- in IOx, we have similar code to effectively merge schemas. This implements compatible definitions of merge.

**Comment by houqp** [comment]
@alamb good call, I only assumed that logically makes sense, but never checked to see if it's actually implemented in datafusion myself. I have pushed a commit to check for fields count in the `contains` method for now. My original use-case was mainly to use a schema with extra metadata for a memtable, which is the case for if schema and record batches are created from our parquet crate.

I am a fan of your SchemaAdapterStream implementation, looks like it would be useful to include the core of that logic in datafusion as well.

File: rust/datafusion/src/datasource/memory.rs [hunk 3]
```diff
@@ -365,4 +366,91 @@ mod tests {
...
+
+        let batch = RecordBatch::try_new(
+            schema1,
+            vec![
+                Arc::new(Int32Array::from(vec![1, 2, 3])),
+                Arc::new(Int32Array::from(vec![7, 5, 9])),
+            ],
+        )?;
+
+        match MemTable::try_new(schema2, vec![vec![batch]]) {
```
Comment (alamb) [inline rust/datafusion/src/datasource/memory.rs:391]: 👍

**Review by alamb — APPROVED** [review]
I think this looks great @houqp  - thank you.  @nevi-me / @jorgecarleitao / @andygrove  any comments?

**Review by alamb — COMMENTED** [review]
Sorry for the delay in merging @houqp  -- it has been a busy week!

**Pull request closed without merging**